### PR TITLE
custom code node

### PR DESCRIPTION
**Pull request by jckhang**

### CODE REVIEW

**Comment by jckhang** [comment]
add pipeline to csvblob and jsonblob as constructor arguments

File: accern_xyme/accern_xyme.py [hunk 1]
```diff
@@ -13,17 +14,20 @@
...
 import time
 import weakref
 import contextlib
 import collections
 from io import BytesIO, StringIO
-import requests
 import pandas as pd
 import quick_server
+import requests
+from RestrictedPython import compile_restricted
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:30]: move it to the function...

File: accern_xyme/accern_xyme.py [hunk 2]
```diff
@@ -1035,6 +1046,81 @@ def get_csv_blob(self) -> 'CSVBlobHandle':
...
+    def set_custom_code(self, func: DF_FUNC) -> CustomCodeResponse:
+        self.check_custom_code_node()
+
+        def as_str(fun: DF_FUNC) -> str:
+            body = inspect.getsource(fun)
+            returns = RETURN_PATTERN.search(body)
+            if returns is None:
+                raise ValueError("no return from custom function")
+            res = f"{body}\nresult = {fun.__name__}(df)"
+            compile_restricted(res, "inline", "exec")
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:1090]: only use if exist

Comment (jckhang) [reply accern_xyme/accern_xyme.py:1090]: and move to util

File: accern_xyme/accern_xyme.py [hunk 3]
```diff
@@ -87,6 +94,10 @@
 INPUT_EXT = [INPUT_ZIP_EXT, INPUT_CSV_EXT, INPUT_TSV_EXT]
 
 
+RETURN_PATTERN = re.compile(r"\s*return\s*")
+DF_FUNC = Callable[[pd.DataFrame], pd.DataFrame]
+
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:99]: add for json, json, 
and json, df

File: accern_xyme/accern_xyme.py [hunk 4]
```diff
@@ -1035,6 +1046,81 @@ def get_csv_blob(self) -> 'CSVBlobHandle':
...
+                f"can not append jsons to {self}, expected 'jsons_reader'")
+        res = cast(JSONBlobResponse, self._client._request_json(
+            METHOD_GET, "/json_blob", {
+                "pipeline": self.get_pipeline().get_id(),
+                "node": self.get_id(),
+            }, capture_err=False))
+        return JSONBlobHandle(self._client, res["json"], res["count"])
+
+    def check_custom_code_node(self) -> None:
+        if not self.get_type().startswith("custom"):
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:1061]: if type in ["custom_json",...]

Comment (jckhang) [reply accern_xyme/accern_xyme.py:1061]: and global variable

File: accern_xyme/accern_xyme.py [hunk 5]
```diff
@@ -1035,6 +1046,81 @@ def get_csv_blob(self) -> 'CSVBlobHandle':
...
+                "pipeline": self.get_pipeline().get_id(),
+                "node": self.get_id(),
+            }, capture_err=False))
+
+    def get_input_example(self) -> Optional[pd.DataFrame]:
+        if self.get_type() != "custom":
+            raise ValueError(
+                "can only load example input data for 'custom' node")
+        res = self._client._request_json(
+            METHOD_GET, "/get_input_example", {
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:1114]: update use /read_node

File: accern_xyme/accern_xyme.py [hunk 6]
```diff
@@ -1179,6 +1265,38 @@ def add_from_df(
 
 
 # *** CSVBlobHandle ***
+# *** JSONBlobHandle ***
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:1268]: remove this line

File: accern_xyme/accern_xyme.py [hunk 7]
```diff
@@ -1179,6 +1265,38 @@ def add_from_df(
...
+            pipeline_id: PipelineHandle) -> 'JSONBlobHandle':
+        res = self._client._request_json(
+            METHOD_PUT, "/json_append", {
+                "pipeline": pipeline_id.get_id(),
+                "blob": self.get_uri(),
+                "jsons": jsons,
+            }, capture_err=True)
+        self._count = res["count"]
+        return self
+
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:1297]: one empy line before and two after

File: requirements.txt [hunk 8]
```diff
@@ -2,4 +2,5 @@ numpy>=1.17.3
 pandas>=0.25.3
 quick-server>=0.7.8
 requests>=2.22.0
+RestrictedPython==5.0
```
Comment (jckhang) [inline requirements.txt:5]: add to setup.py

File: accern_xyme/accern_xyme.py [hunk 9]
```diff
@@ -87,6 +95,16 @@
 INPUT_EXT = [INPUT_ZIP_EXT, INPUT_CSV_EXT, INPUT_TSV_EXT]
 
 
+RETURN_PATTERN = re.compile(r"\s*return\s*")
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:98]: what are those regexes?

File: accern_xyme/accern_xyme.py [hunk 10]
```diff
@@ -625,6 +643,10 @@ def set_job(self, job: JobInfo) -> JobInfo:
                 "job": job,
             }, capture_err=False))
 
+    def get_allowed_custom_imports(self) -> CustomImportsResponse:
+        return cast(CustomImportsResponse, self._request_json(
+            METHOD_GET, "/get_allowed_custom_imports", {}, capture_err=False))
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:648]: rename endpoints without the double encoded method

File: requirements.txt (hunk 8 above)
Comment (JosuaKrause) [inline requirements.txt:5]: `==` -> `>=` match setup.py

File: accern_xyme/accern_xyme.py [hunk 11]
```diff
@@ -1070,7 +1092,104 @@ def get_csv_blob(self) -> 'CSVBlobHandle':
...
+        self.check_custom_code_node()
+
+        def as_str(fun: FUNC) -> str:
+            body = textwrap.dedent(inspect.getsource(fun))
+            returns = RETURN_PATTERN.search(body)
+            if returns is None:
+                raise ValueError("no return from custom function")
+            res = f"{body}\nresult = {fun.__name__}(data)"
+            if PRINT_PATTERN.search(body) is not None:
+                res = f"{res}\nprint_stmt = printed"
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1151]: why not always do this?

File: accern_xyme/accern_xyme.py (hunk 10 above)
Comment (JosuaKrause) [reply accern_xyme/accern_xyme.py:645]: you missed at least this one

**Pull request merged**